pkg/image/unprivileged/options.go

```diff
@@ -0,0 +1,141 @@
+package unprivileged
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+
+	contentlocal "github.com/containerd/containerd/content/local"
+	"github.com/containerd/containerd/metadata"
+	"github.com/containerd/containerd/platforms"
+	"github.com/sirupsen/logrus"
+	bolt "go.etcd.io/bbolt"
+)
+
+type RegistryConfig struct {
+	Log               *logrus.Entry
+	ResolverConfigDir string
+	DBPath            string
+	CacheDir          string
+	PreserveCache     bool
+	SkipTLS           bool
+}
+
+func (r *RegistryConfig) apply(options []RegistryOption) {
+	for _, option := range options {
+		option(r)
+	}
+}
+
+func (r *RegistryConfig) complete() error {
+	if err := os.Mkdir(r.CacheDir, os.ModePerm); err != nil && !os.IsExist(err) {
+		return err
+	}
+
+	if r.DBPath == "" {
+		r.DBPath = filepath.Join(r.CacheDir, "metadata.db")
+	}
+
+	return nil
+}
+
+func defaultConfig() *RegistryConfig {
+	config := &RegistryConfig{
+		Log:               logrus.NewEntry(logrus.New()),
+		ResolverConfigDir: "",
+		CacheDir:          "cache",
+	}
+
+	return config
+}
+
+func NewRegistry(options ...RegistryOption) (*Registry, error) {
+	config := defaultConfig()
+	config.apply(options)
+	if err := config.complete(); err != nil {
+		return nil, err
+	}
+
+	cs, err := contentlocal.NewStore(config.CacheDir)
+	if err != nil {
+		return nil, err
+	}
+
+	bdb, err := bolt.Open(config.DBPath, 0644, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		once   sync.Once
+		closed bool
+	)
+	close := func() error {
+		defer func() {
+			once.Do(func() {
+				closed = true
+			})
+		}()
+		if closed {
```

Comment:
Isn't this read unsafe? If `closed` is set to true in the deferred func, then it's going to be set to true regardless of when `close` returns, so you may as well scrap the defer and replace it with a synchronized get-then-set or an atomic swap at the beginning of `close`.

Reply:
I didn't intend for this to be threadsafe, but you make a good point -- do you think it would suffice to wrap the entire block with `once.Do`?